Update doc of graph api (#505)
Oneflow-Inc/oneflow-documentation: en/docs/cookies/zero.md

2. **Residual States**. It includes activation functions, temporary buffers, and unusable memory fragments.

ZeRO-DP can be divided into three stages, eliminating memory redundancy by partitioning the OPG state rather than copying it directly, and each GPU only saves part of the OPG. Specifically, ZeRO-DP has three main optimization stages, corresponding to O, P, and G respectively. The three stages increase step by step:

> 1. Optimizer states partition（P<sub>os</sub>）: This state is 4x less memory consumption and the same amount of traffic as data parallelism.

Reviewer: ```suggestion
1. Optimizer states partition（P<sub>os</sub>）: This state is 4x less memory consumption and the same amount of traffic as data parallel.
```
这里应该是“数据并行”，在[《大模型训练难于上青天？效率超群、易用的“李白”模型库来了》](https://mp.weixin.qq.com/s/Xrbz0PV_kciAQI61v5JmFQ)中使用的是 data parallel。

> 2. Add gradients partition optimizer (P<sub>os+g</sub>): At this stage, the memory consumption is reduced by 8 times, and the traffic is the same as the data parallelism.

Reviewer: ```suggestion
2. Add gradients partition optimizer (P<sub>os+g</sub>): At this stage, the memory consumption is reduced by 8 times, and the traffic is the same as the data parallel.
```
同17

> 3. Add parameter partition optimizer (P<sub>os+g+p</sub>): At this stage, the memory occupied by the model is evenly distributed among each GPU. Memory consumption is linearly inversely proportional to the degree of data parallelism, but there will be a slight increase in traffic.

Reviewer: ```suggestion
3. Add parameter partition optimizer (P<sub>os+g+p</sub>): At this stage, the memory occupied by the model is evenly distributed among each GPU. Memory consumption is linearly inversely proportional to the degree of data parallel, but there will be a slight increase in traffic.
```
同17

        ...
```

!!! Note
    When using the model for continuous training and prediction: After the training is performed once, ZeRO will automatically change the SBP parameter of the model from Broadcast to Split; when performing prediction, Split will be used for automatic inference without configuring ZeRO.

#### Enable Stage 2 of ZeRO

```python
class CustomGraph(flow.nn.Graph):
    def __init__(self):
        super().__init__()
        ...
        # Enable stage 2 of ZeRO
        self.config.enable_zero(True, stage=2)
        ...
```

Generally speaking, the optimization of stage 2 has large optimization of memory and small speed impact, so it is recommended to use stage 2 optimization. It can be enabled in a simpler way:

```python
class CustomGraph(flow.nn.Graph):
    def __init__(self):
        super().__init__()
        ...
        # Enable stage 2 of ZeRO
        self.config.enable_zero()
        ...
```

#### Enable Stage 3 of ZeRO

```python
class CustomGraph(flow.nn.Graph):
    def __init__(self):
        super().__init__()
        ...
        # Enable stage 3 of ZeRO
        self.config.enable_zero(True, stage=3)
        ...
```

> Although enabling the third stage can minimize the memory consumption, it will increase the communication cost which will lead to lower speed.

Reviewer: ```suggestion
Although enabling the third stage can minimize the memory consumption, it will increase the communication cost which will bring lower speed.
```
